packages/eslint-plugin/tests/rules/no-unnecessary-type-parameters.test.ts

```diff
@@ -0,0 +1,496 @@
+import { RuleTester } from '@typescript-eslint/rule-tester';
+
+import rule from '../../src/rules/no-unnecessary-type-parameters';
+import { getFixturesRootDir } from '../RuleTester';
+
+const jestConsole = console;
+
+beforeEach(() => {
+  global.console = require('console');
+});
+
+afterEach(() => {
+  global.console = jestConsole;
+});
+
+const rootPath = getFixturesRootDir();
+
+const ruleTester = new RuleTester({
+  parser: '@typescript-eslint/parser',
+  parserOptions: {
+    sourceType: 'module',
+    tsconfigRootDir: rootPath,
+    project: './tsconfig.json',
+  },
+});
+
+ruleTester.run('no-unnecessary-type-parameters', rule, {
+  valid: [
+    `
+      class ClassyArray<T> {
+        arr1: T[];
+        arr2: T[];
+      }
+    `,
+    `
+      class ClassyArray<T> {
+        arr: T[];
+        constructor(arr: T[]) {
+          this.arr = arr;
+        }
+      }
+    `,
+    `
+      class ClassyArray<T> {
+        arr: T[];
+        workWith(value: T) {
+          this.arr.indexOf(value);
+        }
+      }
+    `,
+    `
+      abstract class ClassyArray<T> {
+        arr: T[];
+        abstract workWith(value: T): void;
+      }
+    `,
+    `
+      class Box<T> {
+        val: T | null = null;
+        get() {
+          return this.val;
+        }
+      }
+    `,
+    `
+      class Joiner<T extends string | number> {
```

Comment:
Why are the `Joiner` classes considered valid? This one is discussed in the blog post https://effectivetypescript.com/2020/08/12/generics-golden-rule/

Reply:
Oh yeah I was very displeased to 💡 figure this one out. It's that `T[]` expands into several methods that have relation to `T`. Simplifying the test case:

```ts
declare function join(els: { pop: () => T; [i: number]: T; }): void;
```

That `T` is therefore used >1 times. 😬